### PR TITLE
Implement `MDN/send` method/response

### PR DESCRIPTION
**Pull request by sherlockvn**
### Issues

https://github.com/linagora/tmail-flutter/issues/878

### CODE REVIEW

**Review by dab246 — COMMENTED** [review]
### Use `MDN/send`

### Header

```
Accept: application/json; charset=UTF-8; jmapVersion=rfc-8621
```

#### Request

```json
{
    "using": [
        "urn:ietf:params:jmap:mail",
        "urn:ietf:params:jmap:mdn",
        "urn:ietf:params:jmap:core"
    ],
    "methodCalls": [
        [
            "MDN/send",
            {
                "accountId": "{{accountId}}",
                "identityId": "{{identityId}}",
                "send": {
                    "k1546": {
                        "forEmailId": "31ed5fe0-4486-11ed-b412-4700d9090322",
                        "subject": "Subject MDN/send",
                        "textBody": "Hello MDN/send",
                        "disposition": {
                            "actionMode": "manual-action",
                            "sendingMode": "mdn-sent-manually",
                            "type": "displayed"
                        }
                    }
                },
                "onSuccessUpdateEmail": {
                    "#k1546": {
                        "keywords/$mdnsent": true
                    }
                }
            },
            "0"
        ]
    ]
}
```

#### Response

- If the email id matches an existing email message without the `$mdnsent` keyword, the server can answer:

```json
{
    "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
    "methodResponses": [
        [
            "MDN/send",
            {
                "accountId": "0d14dbabe6482aff5cbf922e04cef51a40b4eabccbe12d28fe27c97038752555",
                "sent": {
                    "k1546": {
                        "originalRecipient": "rfc822; firstname23.surname23@upn.integration-open-paas.org",
                        "finalRecipient": "rfc822; firstname23.surname23@upn.integration-open-paas.org",
                        "includeOriginalMessage": false
                    }
                }
            },
            "0"
        ],
        [
            "Email/set",
            {
                "accountId": "0d14dbabe6482aff5cbf922e04cef51a40b4eabccbe12d28fe27c97038752555",
                "oldState": "324b8570-4486-11ed-b412-4700d9090322",
                "newState": "324b8570-4486-11ed-b412-4700d9090322",
                "updated": {
                    "31ed5fe0-4486-11ed-b412-4700d9090322": null
                }
            },
            "0"
        ]
    ]
}
```

 - If the `$mdnsent` keyword has already been set, the server can answer an error:

```json
{
    "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
    "methodResponses": [
        [
            "MDN/send",
            {
                "accountId": "0d14dbabe6482aff5cbf922e04cef51a40b4eabccbe12d28fe27c97038752555",
                "sent": {
                    "k1546": {
                        "originalRecipient": "rfc822; firstname23.surname23@upn.integration-open-paas.org",
                        "finalRecipient": "rfc822; firstname23.surname23@upn.integration-open-paas.org",
                        "includeOriginalMessage": false
                    }
                }
            },
            "0"
        ],
        [
            "Email/set",
            {
                "accountId": "0d14dbabe6482aff5cbf922e04cef51a40b4eabccbe12d28fe27c97038752555",
                "oldState": "324b8570-4486-11ed-b412-4700d9090322",
                "newState": "324b8570-4486-11ed-b412-4700d9090322",
                "updated": {
                    "31ed5fe0-4486-11ed-b412-4700d9090322": null
                }
            },
            "0"
        ]
    ]
}
```

File: lib/jmap/mdn/disposition.dart [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import 'package:equatable/equatable.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'disposition.g.dart';
+
+@JsonSerializable()
+class Disposition with EquatableMixin {
+  late ActionMode actionMode;
```
Comment (dab246) [inline lib/jmap/mdn/disposition.dart:8]: why use `late` without using `final`

File: lib/jmap/mdn/disposition.dart [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  late SendingMode sendingMode;
```
Comment (dab246) [inline lib/jmap/mdn/disposition.dart:9]: same

File: lib/jmap/mdn/disposition.dart [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  late DispositionType type;
```
Comment (dab246) [inline lib/jmap/mdn/disposition.dart:10]: same

File: lib/jmap/mdn/mdn.dart [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  
+  final String? originalRecipient;
+  
+  final String? finalRecipient;
+  
+  final String? originalMessageId;
+  
+  final List<String>? error;
+  
+  Map<String,String>? extensionFields;
```
Comment (dab246) [inline lib/jmap/mdn/mdn.dart:35]: should use `final`

File: lib/jmap/mdn/mdn.dart [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+@JsonSerializable()
+class Mdn with EquatableMixin {
+
+  final Id? forEmailId;
+  
+  final String? subject;
+  
+  final String? textBody;
+  
+  bool includeOriginalMessage;
```
Comment (dab246) [inline lib/jmap/mdn/mdn.dart:19]: should use final

File: lib/jmap/mdn/method/send/send_message_method.dart [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    this.onSuccessUpdateEmail,
+  }) : super(accountId);
+
+  @override
+  MethodName get methodName => MethodName('MDN/send');
+
+  @override
+  Set<CapabilityIdentifier> get requiredCapabilities => {
+        CapabilityIdentifier.jmapMdn,
+        CapabilityIdentifier.jmapMail,
```
Comment (dab246) [inline lib/jmap/mdn/method/send/send_message_method.dart:34]: As the documentation describes `MDN`, we need to add core capability `urn:ietf:params:jmap:core`

File: lib/jmap/mdn/method/send/send_message_response.dart [hunk 7]
```diff
@@ -0,0 +1,52 @@
+import 'package:jmap_dart_client/http/converter/account_id_converter.dart';
+import 'package:jmap_dart_client/http/converter/mdn/mdn_converter.dart';
+import 'package:jmap_dart_client/http/converter/set_error_converter.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/error/set_error.dart';
+import 'package:jmap_dart_client/jmap/core/id.dart';
+import 'package:jmap_dart_client/jmap/core/method/method_response.dart';
+import 'package:jmap_dart_client/jmap/mdn/mdn.dart';
+
+
```
Comment (dab246) [inline lib/jmap/mdn/method/send/send_message_response.dart:10]: remove line

File: lib/jmap/mdn/method/send/send_message_method.dart [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+  @override
+  List<Object?> get props => [
+        methodName,
+        accountId,
+        identityId,
+        send,
+        onSuccessUpdateEmail,
+        requiredCapabilities
+      ];
```
Comment (dab246) [inline lib/jmap/mdn/method/send/send_message_method.dart:67]: remove ident

**Comment by hoangdat** [comment]
still not have test

File: lib/jmap/mdn/method/send/send_message_method.dart (hunk 6 above)
Comment (sherlockvn) [reply lib/jmap/mdn/method/send/send_message_method.dart:34]: i have read the document again, but i don't see where it describe `MDN` need to have `urn:ietf:params:jmap:core`. Can you show me what document is, please? 
image

**Comment by dab246** [comment]
To get `forEmailId` passed in `MDN/send` can be done in 2 ways:
1. Send an email with a header attribute in the outgoing email:
`"header:Disposition-Notification-To:asText": "{{username}}"`

- Request sent
```
"Email/set",
            {
                "accountId": "{{accountId}}",
                "create": {
                    "e01": {
                        "mailboxIds": {
                            "9bc32a50-32cf-11eb-995c-a3ae66e9f96a": true
                        },
                        "subject": "[POSTMAN] SEND EMAIL WITH MDN",
                        "from": [
                            {
                                "name": "{{username}}",
                                "email": "{{username}}"
                            }
                        ],
                        "big": [
                            {
                                "name": "firstname239.surname239",
                                "email": "firstname239.surname239@upn.integration-open-paas.org"
                            }
                        ],
                        "htmlBody": [
                            {
                                "partId": "abc123",
                                "type": "text/html"
                            }
                        ],
                        "bodyValues": {
                            "abc123": {
                                "value": "[POSTMAN] SEND EMAIL WITH MDN",
                                "isEncodingProblem": false,
                                "isTruncated": false
                            }
                        },
                        "header:Disposition-Notification-To:asText": "{{username}}"
                    }
                }
            },
```
- Then get the `id` value in `created` of the `Email/set` method of the returned response.
```
"Email/set",
            {
                "accountId": "0d14dbabe6482aff5cbf922e04cef51a40b4eabccbe12d28fe27c97038752555",
                "oldState": "1bce8420-4551-11ed-a5ca-053c64f285c3",
                "newState": "1bce8420-4551-11ed-a5ca-053c64f285c3",
                "created": {
                    "e01": {
                        "id": "4cb8a430-4551-11ed-96d7-838c1117bed5",
                        "blobId": "4cb8a430-4551-11ed-96d7-838c1117bed5",
                        "threadId": "4cb8a430-4551-11ed-96d7-838c1117bed5",
                        "size": 1011
                    }
                }
            },
            "c1"
```
2. Use `MDN/parse` to get `forEmailId`

File: lib/jmap/mdn/disposition.dart [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  Map<String, dynamic> toJson() => _$DispositionToJson(this);
+}
+
+enum ActionMode {
+  manual('manual-action'),
+  automatic('automatic-action');
+
+  final String value;
+
+  const ActionMode(this.value);
```
Comment (dab246) [inline lib/jmap/mdn/disposition.dart:33]: As the documentation describes, this variable must refer to the values described in the document. So you don't need to create a constructor for it.

Screen Shot 2022-10-07 at 10 42 00

File: lib/jmap/mdn/disposition.dart [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+  }
+}
+
+enum SendingMode {
+  manually('mdn-sent-manually'),
+  automatically('mdn-sent-automatically');
+
+  final String value;
+
+  const SendingMode(this.value);
```
Comment (dab246) [inline lib/jmap/mdn/disposition.dart:52]: same

File: lib/jmap/mdn/disposition.dart [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+enum DispositionType {
+  deleted,
+  dispatched,
+  displayed,
+  processed;
+
+  String toJson() => name;
+
+  static DispositionType fromJson(String json) {
```
Comment (dab246) [inline lib/jmap/mdn/disposition.dart:73]: unnecessary. `enum` will take its own value by the declared name.

File: lib/jmap/mdn/disposition.dart [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  }
+}
+
+enum DispositionType {
+  deleted,
+  dispatched,
+  displayed,
+  processed;
+
+  String toJson() => name;
```
Comment (dab246) [inline lib/jmap/mdn/disposition.dart:71]: unnecessary. `enum` will take its own value by the declared name.

File: lib/jmap/mdn/send/mdn_send_method.dart [hunk 13]
```diff
@@ -1,44 +1,36 @@
...
-import 'package:jmap_dart_client/jmap/core/id.dart';
-import 'package:jmap_dart_client/jmap/core/method/method.dart';
-import 'package:jmap_dart_client/jmap/core/patch_object.dart';
+import 'package:jmap_dart_client/jmap/core/method/request/send_method.dart';
 import 'package:jmap_dart_client/jmap/core/request/request_invocation.dart';
 import 'package:jmap_dart_client/jmap/identities/identity.dart';
+import 'package:jmap_dart_client/jmap/mail/email/submission/set/set_email_submission_method.dart';
 import 'package:jmap_dart_client/jmap/mdn/mdn.dart';
 
-class SendMessageMethod extends MethodRequiringAccountId {
```
Comment (sherlockvn) [inline lib/jmap/mdn/send/mdn_send_method.dart]: What are EmailSubmissionId and ReferenceId field, can you clarify for me, please ?

File: lib/jmap/mdn/send/mdn_send_method.dart [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+import 'package:jmap_dart_client/http/converter/send/send_method_properties_converter.dart';
+import 'package:jmap_dart_client/http/converter/set/set_method_properties_converter.dart';
+import 'package:jmap_dart_client/jmap/core/capability/capability_identifier.dart';
+import 'package:jmap_dart_client/jmap/core/method/request/send_method.dart';
+import 'package:jmap_dart_client/jmap/core/request/request_invocation.dart';
+import 'package:jmap_dart_client/jmap/identities/identity.dart';
+import 'package:jmap_dart_client/jmap/mail/email/submission/set/set_email_submission_method.dart';
+import 'package:jmap_dart_client/jmap/mdn/mdn.dart';
+
+class MDNSendMethod extends SendMethod<MDN> with OptionalOnSuccessUpdateEmail {
```
Comment (sherlockvn) [inline lib/jmap/mdn/send/mdn_send_method.dart:12]: What are ReferenceId and EmailSubmissionId, can you clarify for me, please ?

Comment (dab246) [reply lib/jmap/mdn/send/mdn_send_method.dart:12]: - `ReferenceId` is an id that refers to another id. Defined with the `Prefix` reference keyword and the `Id` value. Example: `#123`
- `EmailSubmissionId` is the id of the `EmailSubmission` method of `Email`

**Review by hoangdat — APPROVED** [review]

**Pull request merged**